src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,416 @@
+            => Math.Max(smallest, Math.Min(value, largest));
+
+            static bool PositionInSnapshot(int position, SourceText text)
+                => GetValueInValidRange(position, 0, Math.Max(0, text.Length - 1)) == position;
+        }
+
+        /// <summary>
+        /// Formats the span between the opening and closing points, options permitting.
+        /// Returns the text changes that should be applied to the input document to 
+        /// get the formatted text and the new closing point in the formatted text.
+        /// </summary>
+        private static async Task<(ImmutableArray<TextChange> TextChanges, int NewClosingPoint)> FormatTrackingSpanAsync(
+            Document document,
+            int openingPoint,
+            int closingPoint,
+            bool shouldHonorAutoFormattingOnCloseBraceOption,
+            ImmutableArray<AbstractFormattingRule>? rules,
+            CancellationToken cancellationToken)
+        {
+            var option = document.Project.Solution.Options.GetOption(BraceCompletionOptions.AutoFormattingOnCloseBrace, document.Project.Language);
+            if (!option && shouldHonorAutoFormattingOnCloseBraceOption)
+            {
+                return (ImmutableArray<TextChange>.Empty, closingPoint);
+            }
+
+            // Annotate the original closing brace so we can find it after formatting.
+            document = await GetDocumentWithAnnotatedClosingBraceAsync(document, closingPoint, cancellationToken).ConfigureAwait(false);
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var startPoint = openingPoint;
+            var endPoint = closingPoint;
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Do not format within the braces if they're on the same line for array/collection/object initializer expressions.
+            // This is a heuristic to prevent brace completion from breaking user expectation/muscle memory in common scenarios.
+            // see bug Devdiv:823958
+            if (text.Lines.GetLineFromPosition(startPoint) == text.Lines.GetLineFromPosition(endPoint))
+            {
+                var startToken = root.FindToken(startPoint, findInsideTrivia: true);
+                if (startToken.IsKind(SyntaxKind.OpenBraceToken) &&
+                    (startToken.Parent?.IsInitializerForArrayOrCollectionCreationExpression() == true ||
+                     startToken.Parent is AnonymousObjectCreationExpressionSyntax))
+                {
+                    // format everything but the brace pair.
+                    var endToken = root.FindToken(endPoint, findInsideTrivia: true);
+                    if (endToken.IsKind(SyntaxKind.CloseBraceToken))
+                    {
+                        endPoint -= endToken.Span.Length + startToken.Span.Length;
+                    }
+                }
+            }
```

Comment:
i don't really understand thse comments.  the first comment says "do not format wihin the braces.  but hte second comment says "format everything but the braces...

i don't get it.  nor do i even understand the logic.  if we're formatting eveyrthing but hte brace pair... shouldn't the start point move forward, and the endpoint backward?